lib/checkclass.cpp

```diff
@@ -2316,20 +2316,31 @@ bool CheckClass::checkConstFunc(const Scope *scope, const Function *func, bool&
                 return false;
             memberAccessed = true;
         }
-        bool mayModifyArgs = true;
-        if (const Function* f = funcTok->function()) { // TODO: improve (we bail out if there is any possible modification of any argument)
+
+        if (const Function* f = funcTok->function()) { // check known function
             const std::vector<const Token*> args = getArguments(funcTok);
             const auto argMax = std::min<nonneg int>(args.size(), f->argCount());
-            mayModifyArgs = false;
+
             for (nonneg int argIndex = 0; argIndex < argMax; ++argIndex) {
                 const Variable* const argVar = f->getArgumentVar(argIndex);
-                if (!argVar || ((argVar->isArrayOrPointer() || argVar->isReference()) && !argVar->isConst())) {
-                    mayModifyArgs = true;
-                    break;
+                if (!argVar || ((argVar->isArrayOrPointer() || argVar->isReference()) &&
+                                !(argVar->valueType() && argVar->valueType()->isConst(argVar->valueType()->pointer)))) { // argument might be modified
```

Comment:
I wonder if `argVar->valueType()->isConst(argVar->valueType()->pointer)))` is strict enough. If pointer is 2 and constness is 0,1,2 the data can be modified

Reply:
It seems that we still have FNs (see added tests), but I couldn't generate any FPs so far. The check `!var->isConst()` below is probably too strict.